src/lib/System/IO/Streams/Heartbeat.hs

```diff
@@ -79,17 +114,25 @@ heartbeatInputStream interval graceMultiplier is = do
 -- the heartbeat interval and the current timestamp, calculate if a
 -- heartbeat must be sent and how much time there is until the next heartbeat
 -- must be sent.
-heartbeatTime :: DiffTime -- ^ Maximum time since last message, ie. heartbeat interval or grace period
-              -> UTCTime  -- ^ Current timestamp
-              -> UTCTime  -- ^ Last message timestamp
-              -> (UTCTime, (DiffTime, Bool)) -- ^ (New last message timestamp, (time til heartbeat, send new message?))
+heartbeatTime ::
+    -- | Maximum time since last message, ie. heartbeat interval or grace period
+    Maybe DiffTime ->
+    -- | Current timestamp
+    UTCTime ->
+    -- | Last message timestamp
+    UTCTime ->
+    -- | (New last message timestamp, (time til heartbeat, send new message?))
+    (UTCTime, (Maybe DiffTime, Bool))
 heartbeatTime interval now lastTime = (if triggerHeartbeat then now else lastTime, (timeTilHeartbeat, triggerHeartbeat))
   where
     timeSinceMsg = realToFrac $ diffUTCTime now lastTime
-    triggerHeartbeat = timeSinceMsg >= interval
-    timeTilHeartbeat = interval - timeSinceMsg
+    triggerHeartbeat = maybe False (timeSinceMsg >=) interval
+    timeTilHeartbeat = fmap (\i -> i - timeSinceMsg) interval
 
 
-delayDiffTime :: DiffTime -> IO ()
-delayDiffTime = threadDelay . picosToMicros
-  where picosToMicros = fromIntegral . diffTimeToPicoseconds . (/ 1000000)
+-- currently this will cause 0 delay if the interval is 'Nothing', maybe we want to have a more sane
+-- default that won't spin so much
+delayDiffTime :: Maybe DiffTime -> IO ()
+delayDiffTime = traverse_ (threadDelay . picosToMicros)
```

Comment:
Hmmmmmmmmm. This seems wrong. We should probably setup a callback on any changes to the interval.

Probably the right way to do this would be to kill the thread if you set the interval to `Nothing`, and then start a new thread when the interval is set again.